data_steward/retraction/retract_deactivated_pids.py

```diff
@@ -0,0 +1,608 @@
+# Python imports
+import re
+import argparse
+import logging
+import os
+from datetime import datetime
+
+# Third party imports
+import pandas as pd
+from jinja2 import Environment
+from google.cloud import bigquery
+
+# Project imports
+from retraction.retract_utils import DEID_REGEX
+from constants import bq_utils as bq_consts
+from constants.cdr_cleaner import clean_cdr as clean_consts
+from sandbox import check_and_create_sandbox_dataset
+
+LOGGER = logging.getLogger(__name__)
+LOGS_PATH = '../logs'
+
+CLIENT = None
+
+
+def get_client(project_id):
+    """
+    Ensure only one client is created and reused
+
+    :param project_id:  project to get a client for
+    :returns: a big query client object
+    """
+    global CLIENT
+    if not CLIENT:
+        CLIENT = bigquery.Client(project=project_id)
+
+    return CLIENT
+
+
+jinja_env = Environment(
+    # help protect against cross-site scripting vulnerabilities
+    autoescape=True,
+    # block tags on their own lines
+    # will not cause extra white space
+    trim_blocks=True,
+    lstrip_blocks=True,
+    # syntax highlighting should be better
+    # with these comment delimiters
+    comment_start_string='--',
+    comment_end_string=' --')
+
+TABLE_INFORMATION_SCHEMA = jinja_env.from_string("""
+SELECT *
+FROM `{{project}}.{{dataset}}.INFORMATION_SCHEMA.COLUMNS`
+""")
+
+DEACTIVATED_PIDS_QUERY = jinja_env.from_string("""
+SELECT DISTINCT *
+FROM `{{project}}.{{dataset}}.{{table}}`
+""")
+
+RESEARCH_ID_QUERY = jinja_env.from_string("""
+SELECT DISTINCT research_id
+FROM `{{project}}.{{prefix_regex}}_combined._deid_map`
+WHERE person_id = {{pid}}
+""")
+
+CHECK_PID_EXIST_DATE_QUERY = jinja_env.from_string("""
+SELECT
+COALESCE(COUNT(*), 0) AS count
+FROM `{{project}}.{{dataset}}.{{table}}`
+WHERE person_id IN (SELECT person_id
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`)
+AND {{date_column}} > (SELECT MIN(deactivated_date)
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`)
+""")
```

Comment:
So, this is checking that a person exists in the deactivated table and a date field exists that is greater then the minimum deactivated date in the deactivation table.  Even if the minimum deactivation date belongs to a different person id.  Is this what you intend?

Reply:
Correct! It's a generalized select to try and eliminate tables that queries need to be created for. If I went through every deactivated date per PID it would be a ton more queries and take a lot longer.